### PR TITLE
Css variables part 1

### PR DESCRIPTION
**Pull request by JNaftali**
A much less ambitious version of https://github.com/wagtail/wagtail/pull/6349

This PR:
* Adds sass functions and mixins for easier authoring/modifying of CSS variables, as documented in the other MR
* Adds `postcss-custom-properties` to the sass gulp task so that the admin will continue working on IE despite the lack of support for css variables
* changes the `$color-teal` sass variable (and all the variables that depend on it) to reference the `--color-primary` css variable

### CODE REVIEW

**Comment by JNaftali** [comment]
@thibaudcolas As far as your comment:
> It feels to me like it would be better to just refactor existing colors to not rely on modifying the darkness/lightness/saturation

How would you recommend dealing with the one-off colors in _listing.scss and _tabs.scss? Add a new color variable (`$color-teal-darkerer`? `$color-teal-darkest`?) that's close enough to those values that I can replace them?

**Review by allcaps — CHANGES_REQUESTED** [review]
@JNaftali Thank you for the PR. Great effort! Also on the more advanced one :clap:

I've been requested to review your PR. Maybe because I contributed the previous color changes. When we went from low contrast lime to the higher contrast teal.

How does this work in practice? What does an integrator need to know to change the admin colours? The docs are missing. I need the guidance too ;)

I do get that branding is important. More important is that Wagtail strives to be WCAG2.1, AA level compatible. The Wagtail color scheme and contrasts ratios are carefully crafted to comply to Web Content Accessibility Guidelines. WCAG 2.2 AA. I'd like the docs to have a warning about braking contrast and how to avoid that.

**Comment by JNaftali** [comment]
> @JNaftali Thank you for the PR. Great effort! Also on the more advanced one 👏
> 
> I've been requested to review your PR. Maybe because I contributed the previous color changes. When we went from low contrast lime to the higher contrast teal.

Thanks for the review!

> How does this work in practice? What does an integrator need to know to change the admin colours? The docs are missing. I need the guidance too ;)

I had been holding off on docs while I wasn't sure if the PR was going to go through any major changes. I'll have a go at writing some tonight or tomorrow.

In the mean time - to customize the admin colors you need to insert a stylesheet into the wagtail admin (often using https://docs.wagtail.io/en/v2.10.2/reference/hooks.html#insert-global-admin-css ). That stylesheet needs to have a :root selector that defines some css variables. For example:

```css
:root {
  --color-primary-hue: 45;
}
```

This PR adds 5 customizable colors: `color-primary`, `color-primary-darker`, `color-primary-dark`, `color-input-focus`, and `color-input-focus-border`. Ideally these should be set using 3 css variables per color - `color-primary-hue` (a unitless number), `color-primary-saturation` (percentage), and `color-primary-lightness` (percentage) (with equivalents for each of the other 4 colors). Assuming you choose a good color for `color-primary`, you shouldn't have to touch any of the other colors.

> I do get that branding is important. More important is that Wagtail strives to be WCAG2.1, AA level compatible. The Wagtail color scheme and contrasts ratios are carefully crafted. I'd like the docs to have a warning about braking contrast and how to avoid that.

That was absolutely something that I planned to do. Do you have a preferred browser extension for checking contrast values? If not I'll probably call out Axe.

Will try to update the PR with docs tonight

File: client/scss/tools/_various.colors.scss
Comment (thibaudcolas) [inline client/scss/tools/_various.colors.scss:31]: Is this and the other HSL shifts really equivalent to the Sass color functions they replace?

**Review by thibaudcolas — COMMENTED** [review]
Looking as amazing as the last PR @JNaftali. There is one thing I’d like to confirm – this approach relies on HSL, and as-is it wouldn’t be possible for people to only set `color-primary` in RGB hex and have it trickle down to the other shades?

Can you confirm that the cases below do behave as I’d expect?

```scss
    :root {
        // As you documented – this will auto-magically adjust all dark, darker, input, input border shades.
        --color-primary-hue: 25;

        // This one correctly overrides the primary color, but the dark and darker variations will still be based on our teal
        // since they are defined based on the -hue, -saturation, -lightness variables.
        --color-primary: #ff0000;
    }
```

So practically, for real-world theming, people will either have to know which primary color they want to use in HSL notation, and know how to split a normal HSL value into the three components:

```scss
:root {
    // This will trickle down to the dark and darker shades as expected.
    --color-primary-hue: 25;
    --color-primary-saturation: 25;
    --color-primary-lightness: 25;
}
```

…or if they want to use RGB / hex values, then they have to override _all_ of the colors:

```scss
:root {
    --color-primary: #ff0000;
    --color-primary-dark: #cc0000;
    --color-primary-darker: #990000;
    --color-input-focus: #dd00f0;
    --color-input-focus-border: #dd00ff;
}
```

Is that correct? I’m still uncertain of whether the HSL trick really is worthwhile or not.

File: client/scss/tools/_various.colors.scss
Comment (JNaftali) [reply client/scss/tools/_various.colors.scss:31]: Near as I can tell, yep! Run the following through sass (I did it on sassmeister.com to check easily) and you'll see the two colors end up identical:
```scss
.test {
    color: darken(red, 10%);
    background-color: hsl(hue(red), saturation(red), lightness(red) - 10%);
}
```

**Comment by JNaftali** [comment]
Yup, everything you wrote there is correct. I don't know of any work coming up to standardize anything like this - since coming up with all this I've run across https://una.im/css-color-theming/ so I know that I'm not the only one to imagine doing stuff this way. It wouldn't shock me if there ends up being a way to make color functions in css whenever houdini support becomes the norm but I haven't seen anything written about that specifically.

Seems like even sass is trying to de-emphasize darken in favor of `color.scale` - https://sass-lang.com/documentation/modules/color#darken

**Comment by JNaftali** [comment]
@allcaps Hey! Hate to be a bother, but are the docs that I added sufficient? Or are there further changes you would like to request?

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (allcaps) [inline docs/advanced_topics/customisation/admin_templates.rst:97]: `UI` => `user interface`. No abbrs plz.

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (allcaps) [inline docs/advanced_topics/customisation/admin_templates.rst:105]: H of HSL is a 360' number of the color wheel starting at red.

`(a unitless number)` => `(0-360 without unit)`

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (allcaps) [inline docs/advanced_topics/customisation/admin_templates.rst:110]: Please use the warning directive here.

```
Warnings
---------
```
=>
```
.. warning::
    This is warning text. 
```
Just use two times the warning directive. No numbering.

**Review by allcaps — APPROVED** [review]
Sorry to let you hanging. Thank you for your patience.

Overall it looks great. Just some nitpicks.

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (allcaps) [reply docs/advanced_topics/customisation/admin_templates.rst:97]: `css` => `CSS`

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (allcaps) [reply docs/advanced_topics/customisation/admin_templates.rst:105]: `css` => `CSS`

**Comment by thibaudcolas** [comment]
Thank you @JNaftali, honestly this is so future-thinking I wasn’t too sure how to review it. I wasn’t expecting something so fancy! I’ll do one more round of tests on Friday, and hopefully merge this then.

**Comment by JNaftali** [comment]
Thanks for the review @allcaps! Made those changes.

And that's super exciting to hear, @thibaudcolas! This is my first time contributing a feature to an open source project that I use, nice to hear it described like that

**Comment by allcaps** [comment]
Again, thank you! This is a much requested feature. I'm sure many people are happy to see it realised. :D 

I'll leave the final review and merge to @thibaudcolas.

File: client/scss/settings/_variables.scss
Comment (thibaudcolas) [inline client/scss/settings/_variables.scss:34]: I would welcome a follow-up PR to search-and-replace all the Sass variables so they switch to `color-primary` as well. But glad this isn’t in the same PR for now!

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
This looks great in Chrome, unfortunately it’s broken in IE11 as far as I can see. The fallback for the primary color is there, but the `calc` fails for the darker shades. Did you do any testing there?

ie11-calc

I’ll have a look over the weekend or next week to understand what the issue is.

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (thibaudcolas) [inline docs/advanced_topics/customisation/admin_templates.rst:128]: I think all of these should be the first thing in this section, not last. And combined in a single warning?

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (thibaudcolas) [inline docs/advanced_topics/customisation/admin_templates.rst:105]: This all sounds very advanced – I think we’re missing an intermediary example of "here is how you can set all the colors manually with hex codes".

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (JNaftali) [reply docs/advanced_topics/customisation/admin_templates.rst]: done

File: docs/advanced_topics/customisation/admin_templates.rst
Comment (JNaftali) [reply docs/advanced_topics/customisation/admin_templates.rst:111]: added such an example to the end

**Comment by JNaftali** [comment]
@thibaudcolas I did not actually test on IE11 - I don't have a great way to run it, so your help debugging that would be very much appreciated.

**Comment by thibaudcolas** [comment]
According to [caniuse](https://caniuse.com/calc) it looks like IE11 doesn’t support `calc` in color functions. No matter, the usage of `calc` in the IE11-only fallback really is just a remnant of the custom properties setup and shouldn’t actually be needed. I’ll try to use https://github.com/postcss/postcss-calc to remove the unneeded & incompatible `calc`s and report back.

File: client/scss/settings/_variables.scss
Comment (thibaudcolas) [inline client/scss/settings/_variables.scss:64]: Looks like this should be:

```suggestion
$color-input-focus-border: var(--color-input-focus-border);
```

File: client/scss/elements/_root.scss
Comment (thibaudcolas) [inline client/scss/elements/_root.scss:8]: Based on the previous Sass definition this should be:

```suggestion
    @include define-color('color-input-focus-border', css-lighten(css-saturate(get-color('color-primary'), 12%), 10%));
```

**Review by thibaudcolas — CHANGES_REQUESTED** [review]

**Comment by JNaftali** [comment]
@thibaudcolas Still haven't managed to get IE running to test there but I added postcss-calc as you suggested and it seems to have removed calc from the backup declarations. So hopefully fixed whatever issues existed on IE

**Comment by thibaudcolas** [comment]
👍  thank you, I’ll assign this back to myself and see whether that was enough / take over from there.

**Comment by dest81** [comment]
Thank you that doing this @JNaftali! I have played with your branch and  it works very well. 
So far I found one thing - in the stream field add block button on hover has old default wagtail color #007d7e. The problem is [here](https://github.com/wagtail/react-streamfield/blob/master/src/scss/_variables.scss#L19).

**Comment by JNaftali** [comment]
> So far I found one thing - in the stream field add block button on hover has old default wagtail color #007d7e. The problem is [here](https://github.com/wagtail/react-streamfield/blob/master/src/scss/_variables.scss#L19).

Thanks for testing this out!, @dest81!

Given that react-streamfield is a separate package I don't think I want to mess with that until this PR is merged in... But! Since css variables exist at runtime, react-streamfield could fairly easily reference wagtail's color variables in their css. Thanks for making me aware of this, I might take a swing at that next :)

**Comment by dest81** [comment]
@thibaudcolas I wonder if/when this PR can be merged?

**Comment by thibaudcolas** [comment]
@dest81 when I get time to test it in IE11 and it works there as well – last time I tested this the styles were completely broken in that browser.

**Comment by JNaftali** [comment]
@thibaudcolas Obviously you should test this in your own time, but I managed to get VirtualBox up and running on my machine and it seems to be running well on IE! Screenshot attached (of IE accessing the Squash branch).

Screen Shot 2020-12-12 at 2 08 11 PM

**Comment by dest81** [comment]
I have tested on IE11 via BrowserStack and can confirm it works.

**Comment by dest81** [comment]
Hi guys,
Sorry that annoying you but any news here? 
I still see "needs work" label on this PR, shouldn't be changed to "Needs review"?
I know your are all busy but can you give me rough estimate when it can be reviewed and possibly merged?

**Comment by thibaudcolas** [comment]
Hey @dest81, you’re right, I’ve updated the labels. I have the branch checked out locally and am halfway through reviewing it. There is a blocker with https://github.com/wagtail/react-streamfield/pull/14 but I managed to find another way to do this that doesn’t involve changes to this separate package, so we’re in a good place. I’ll close that PR once I’ve confirmed my approach does work.

I still have more testing to do – I wouldn’t have noticed https://github.com/wagtail/react-streamfield/pull/14 had @JNaftali not reported there, so I want to make sure I’m not missing other things like this.

One other thing I’m still unsure of is `color-primary-darkest` – it’s very annoying to have to introduce this new shade for just 2 places out of the whole of Wagtail, and I think I found a way to change how those styles are made so we don’t need this. But here as well, more testing to do.

**Comment by dest81** [comment]
@thibaudcolas thank you very much for the quick response, if you need any help (maybe with testing) I would be very glad to help ...

**Comment by JNaftali** [comment]
> One other thing I’m still unsure of is `color-primary-darkest` – it’s very annoying to have to introduce this new shade for just 2 places out of the whole of Wagtail, and I think I found a way to change how those styles are made so we don’t need this. But here as well, more testing to do.

I look forward to hearing what you come up with! Let me know if there's some way I can help

**Review by thibaudcolas — APPROVED** [review]
All working as expected across the browsers I tested! And yep, the `darkest` variant shouldn’t be needed in the end.

Thank you @JNaftali for making this happen! This took a fair amount of back and forth but the end result really is remarkable, and considering how frequently this had been requested I’m sure lots of people will find this useful. I’ll follow up in #2943 for any further work in this area.

**Comment by thibaudcolas** [comment]
@dest81 this is now published in Wagtail 2.12rc1 if you want to try it out: https://pypi.org/project/wagtail/2.12rc1/. Release notes: https://docs.wagtail.io/en/latest/releases/2.12.html#admin-color-themes.

**Comment by dest81** [comment]
@thibaudcolas I did a quick test and it works as expected. Will work with it more closely after wagtail 2.12 release. Thank you  @JNaftali and @thibaudcolas for the great job

**Pull request merged**
